source/strings.tex

```diff
@@ -4174,6 +4180,36 @@
 and initializing \tcode{size_} with \tcode{len}.
 \end{itemdescr}
 
+\indexlibraryctor{basic_string_view}%
+\begin{itemdecl}
+template<class It, class End>
+  constexpr basic_string_view(It begin, End end);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\constraints
+\begin{itemize}
+\item \tcode{It} satisfies \libconcept{contiguous_iterator}.
+\item \tcode{End} satisfies \tcode{\libconcept{sized_sentinel_for}<It>}.
+\item \tcode{is_same_v<iter_value_t<It>, charT>} is \tcode{true}.
+\item \tcode{is_convertible_v<End, size_type>} is \tcode{false}.
+\end{itemize}
+
+\pnum
+\expects
+\begin{itemize}
+\item \range{begin}{end} is a valid range.
+\item \tcode{It} models \libconcept{contiguous_iterator}.
+\item \tcode{End} models \tcode{\libconcept{sized_sentinel_for}<It>}.
+\end{itemize}
+
+\pnum
+\effects
+Initializes \tcode{data_} with \tcode{to_address(begin)} and
+initializes \tcode{size_} with \tcode{end - begin}.
```

Comment:
`sIze_t` is of type size_type, but `end-begin` is of type `ptrdiff_t`. The conversion will "just work" even for the overflow case, but it's a bit odd.

Reply:
Bug: `end - begin` has some _`signed-integer-like`_ type (that isn't necessarily `ptrdiff_t`) but must be explicitly convertible to `size_t` per [iterator.concept.winc]/6. This needs to initialize `size_` with `static_cast<size_t>(end - begin)`.

As always, let-me-know-if-this-isn't-editorial-and-I'll-file-an-LWG-issue.

Reply:
We could simply say "direct-initializes `size_t` with `end-begin`", but that's a bit subtle.
@zygoloid , your call. (This also affects the similar text in the corresponding `span` changes.)

Reply:
Sorry, my "bug" cry here was a bit of an overreaction - I (and apparently all of LWG) am still not used to non-integral difference types. Either "direct-initializes" or `static_cast<size_t>` would be great here, or we could clean it up post-merge via LWG issue. This certainly doesn't rise to the level of refusing to apply the proposal and sending it back to LWG, which "bug" would seem to imply.

Reply:
It seems to me that this is just one instance of a systemic problem in the library wording. This phrasing "initializes x with y" is common, but underspecified (and formally meaningless) -- the library wording either needs to say what kind of initialization is performed, or specify an *initializer* (not an *expression*) with which to initialize. We should ask LWG to think about this; for each "initializes x with y" utterance, the reader should know what kind of initialization we mean.

Looking at random through the library wording, the first case I found: [queue.cons]/1:

> *Effects:* Initializes `c` with `cont`.

The meaning of this depends on whether this is direct- or copy-initialization. (It's obscure, but if `T` is not *Cpp17CopyInsertable* into the container, it could be the case that one form of initialization works and the other does not, or that they both work and do different things.)

Another random sample: [partial.sum]/2:

>  *Effects:* For a non-empty range, the function creates an accumulator `acc` whose type is `InputIterator`'s value type, initializes it with `*first`, and assigns the result to `*result`.

Again the difference between direct- and copy-initialization is observable here.

Perhaps the library should have blanket wording that when it says "initializes", it means by direct- or copy-initialization, and that it's unspecified which one you get (or something like that) -- and someone should go through all the instances and check if any of them mean something else (I doubt this is the only case that does). I'm not convinced that this particular instance of the problem needs to be dealt with while applying the motions.

@CaseyCarter Can you file an LWG issue, please?

Reply:
I've mailed lwgchair to ask for an issue.